rzscontrol/man/rzscontrol.1
'\" t
.\"     Title: rzscontrol
.\"    Author: Nitin Gupta <ngupta@vflare.org>
.\" Generator: DocBook XSL Stylesheets v1.74.3 <http://docbook.sf.net/>
.\"      Date: 07/19/2009
.\"    Manual:  
.\"    Source:    
.\"  Language: English
.\"
.TH "RZSCONTROL" "1" "07/19/2009" "" ""
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
rzscontrol \- control ramzswap devices
.SH "SYNOPSIS"
.HP \w'\fBrzscontrol\fR\ 'u
\fBrzscontrol\fR {ramzswap\ device} [\-b|\-\-backing_swap=<path>] [\-m|\-\-memlimit_kb=<value>] [\-d|\-\-disksize_kb=<value>] [\-i|\-\-init] [\-r|\-\-reset] [\-s|\-\-stats] [\-v|\-\-verbose] [\-h|\-\-help]
.SH "DESCRIPTION"
.PP
The ramzswap kernel module creates multiple ramzswap devices, equal to NUM_DEVICES parameter\&. With rzscontrol, you can control each of these devices\&.
.PP
This is how you use a ramzswap device (/dev/ramzswapX):
.PP
\- Set parameters for [backing swap, memory limit] or [disk size] (see OPTIONS and EXAMPLES)\&.
.PP
\- Issue init once these parameters are set: rzscontrol /dev/ramzswapX \-\-init\&.
.PP
\- swapon /dev/ramzswapX (where X is ramzswap device id: 0, 1, 2\&.\&.\&. NUM_DEVICES\-1)\&.
.PP
\- When you are done with this swap: swapoff /dev/ramzswapX\&. NOTE: swapoff will not free any compressed pages\&.
.PP
\- After swapoff, issue reset: rzscontrol /dev/ramzswapX \-\-reset\&. This will free all metadata and compressed pages\&.
.SH "OPTIONS"
.PP
\fB\-b, \-\-backing_swap\fR
.RS 4
Specify backing swap device\&. This can either be a swap block device or a swap file\&. Writes are forwarded to this device when memory limit is reached or when a page is incompressible\&. You must swapoff backing swap device before init is issued, otherwise init will fail with \-EBUSY error\&. Also, backing swap must be a valid swap device, otherwise \'swapon /dev/ramzswapX\' will fail as if swapon was issued directly on the backing swap device\&.
.RE
.PP
\fB\-m, \-\-memlimit_kb\fR
.RS 4
Specify memory limit (in KB)\&. This is the limit on amount of memory allocated for compressed pages\&. When this memory limit is reached, all further writes are forwarded to backing swap device\&. This option is valid only if backing swap is specified\&. This must be smaller than or equal to backing swap size\&. Default: 15% of RAM or backing swap size, whichever is smaller\&.
.RE
.PP
\fB\-d, \-\-disksize_kb\fR
.RS 4
Specify ramzswap disk size (in KB)\&. This is the limit on number of (uncompressed) pages that can be stored in this device\&. This parameter is valid only when backing swap is not present since in that case, it is implicitly equal to size of the backing swap device\&. Default: 25% of RAM
.RE
.PP
\fB\-i, \-\-init\fR
.RS 4
Initialize given ramzswap device\&. It causes allocation of metadata for given ramzswap device and makes it ready to receive swap requests\&. Use
\fBswapon\fR(8)
to start using this device as swap\&.
.RE
.PP
\fB\-r, \-\-reset\fR
.RS 4
Resets given ramzswap device\&. It causes deallocation of metadata and any compressed pages for given ramzswap device\&. If the device is currently active, use
\fBswapoff\fR(8)
before issuing reset (otherwise \-EBUSY error is returned)\&.
.RE
.PP
\fB\-s, \-\-stats\fR
.RS 4
Show stats for given ramzswap device\&. If the ramzswap kernel module was compiled without STATS support, then most of the stats will be shown as 0\&.
.RE
.PP
\fB\-v, \-\-verbose\fR
.RS 4
Show verbose output from rzscontrol\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Shows usage information and some examples\&.
.RE
.SH "EXAMPLES"
.PP
Following shows a typical sequence of steps for using ramzswap\&. Initialization can be done with many variations of configuration parameters as shown below\&.
.PP
Load Module:
.sp
.if n \{\
.RS 4
.\}
.nf
insmod ramzswap\&.ko NUM_DEVICES=4 # creates 4 uninitialized devices: /dev/ramzswap{0,1,2,3}
.fi
.if n \{\
.RE
.\}
.PP
Initialize:
.sp
.if n \{\
.RS 4
.\}
.nf
    rzscontrol /dev/ramzswap0 \-\-init # uses default value of disksize_kb
    rzscontrol /dev/ramzswap0 \-\-disksize_kb=10240 \-\-init
    rzscontrol /dev/ramzswap1 \-\-backing_swap=/dev/sda2 \-\-init # uses default value of memlimit_kb
    rzscontrol /dev/ramzswap1 \-\-backing_swap=/dev/sda2 \-\-memlimit_kb=10240 \-\-init
    rzscontrol /dev/ramzswap2 \-\-backing_swap=/path/to/swap\&.file \-\-memlimit_kb=10240 \-\-init
    
.fi
.if n \{\
.RE
.\}
.PP
Activate:
.sp
.if n \{\
.RS 4
.\}
.nf
swapon /dev/ramzswap2 # or any other initialized ramzswap device
.fi
.if n \{\
.RE
.\}
.PP
Stats:
.sp
.if n \{\
.RS 4
.\}
.nf
rzscontrol /dev/ramzswap2 \-\-stats
.fi
.if n \{\
.RE
.\}
.PP
Deactivate:
.sp
.if n \{\
.RS 4
.\}
.nf
swapoff /dev/ramzswap2
.fi
.if n \{\
.RE
.\}
.PP
Reset:
.sp
.if n \{\
.RS 4
.\}
.nf
rzscontrol /dev/ramzswap2 \-\-reset
.fi
.if n \{\
.RE
.\}
.PP
Unload Module:
.sp
.if n \{\
.RS 4
.\}
.nf
rmmod ramzswap\&.ko
.fi
.if n \{\
.RE
.\}
.PP
NOTE: ramzswap parameters cannot be changed once the device is initialized\&. It has to be reset before new parameters can be used\&.
.SH "SEE ALSO"
.PP

\fBmkswap\fR(8),
\fBswapon\fR(8),
\fBswapoff\fR(8)
.SH "AUTHOR"
.PP
\fBNitin Gupta\fR <\&ngupta@vflare\&.org\&>
.RS 4
Developer
.RE


rzscontrol/rzscontrol.c
/*
 * rzscontrol - Control ramzswap devices
 *
 * Copyright (C) 2008, 2009  Nitin Gupta
 *
 * This code is released using a dual license strategy: BSD/GPL
 * You can choose the licence that better fits your requirements.
 *
 * Released under the terms of 3-clause BSD License
 * Released under the terms of GNU General Public License Version 2.0
 *
 * Project home: http://compcache.googlecode.com
 */

#include <errno.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <inttypes.h>

#include "types.h"
#include "ramzswap_ioctl.h"

/* Flag set by '--verbose' */
static int verbose_flag;

/* /dev/ramzswapX */
#define MAX_DEVICE_NAME_LEN	32

/*
 * 10 decimal digits can represent 2^32.
 * All input sizes are in KB, so 10 digits are sufficient.
 */
#define MAX_SIZE_LEN	10

#define VERBOSE(args...) do {	\
	if (verbose_flag)	\
		printf(args);	\
} while (0)

struct rzs_args {
	char backing_swap[MAX_SWAP_NAME_LEN];
	u64 memlimit_kb;
	u64 disksize_kb;
	int init;
	int reset;
	int stats;
};

void usage(void)
{
	printf("Usage: rzscontrol {<ramzswap device>} [<options>]\n"
		"Example: rzscontrol /dev/ramzswap0 --init\n"
		"See rzscontrol manpage for details.\n");
}

void show_stats(struct ramzswap_ioctl_stats *s)
{
	int backing_swap_present = (s->backing_swap_name[0] != '\0');

	if (backing_swap_present)
		printf("BackingSwap:	%s\n", s->backing_swap_name);

#define K(x)	((x) >> 10)
	/* Basic stats */
	printf(
		"DiskSize:	%8" PRIu64 " kB\n",
		K(s->disksize)
	);

	/* memlimit is valid only if backing swap is present */
	if (backing_swap_present)
		printf("MemLimit:	%8" PRIu64 " kB\n", K(s->memlimit));

	/* Extended stats */
	printf(
		"NumReads:	%8" PRIu64 "\n"
		"NumWrites:	%8" PRIu64 "\n"
		"FailedReads:	%8" PRIu64 "\n"
		"FailedWrites:	%8" PRIu64 "\n"
		"InvalidIO:	%8" PRIu64 "\n"
		"NotifyFree:	%8" PRIu64 "\n"
		"ZeroPages:	%8u\n"
		"GoodCompress:	%8u %%\n"
		"NoCompress:	%8u %%\n"
		"PagesStored:	%8u\n"
		"PagesUsed:	%8u\n"
		"OrigDataSize:	%8" PRIu64 " kB\n"
		"ComprDataSize:	%8" PRIu64 " kB\n"
		"MemUsedTotal:	%8" PRIu64 " kB\n",
		s->num_reads,
		s->num_writes,
		s->failed_reads,
		s->failed_writes,
		s->invalid_io,
		s->notify_free,
		s->pages_zero,
		s->good_compress_pct,
		s->pages_expand_pct,
		s->pages_stored,
		s->pages_used,
		K(s->orig_data_size),
		K(s->compr_data_size),
		K(s->mem_used_total)
	);

	if (backing_swap_present) {
		printf(
			"BDevNumReads:	%8" PRIu64 "\n"
			"BDevNumWrites:	%8" PRIu64 "\n",
			s->bdev_num_reads,
			s->bdev_num_writes
		);
	}
}

int do_ioctl(int fd, int argc, struct rzs_args *args)
{
	int ret = 0;

/* Print an error message if ioctl fails */
#define ON_ERR(str) do {				\
	int err;					\
	if (ret == -1) {				\
		err = errno;				\
		printf("%s: %s\n", str, strerror(err));	\
		goto out;				\
	}						\
} while (0)

	while (argc--) {

	if (args->backing_swap[0] != '\0') {
		VERBOSE("backing_swap: %s\n", args->backing_swap);
		ret = ioctl(fd, RZSIO_SET_BACKING_SWAP, args->backing_swap);
		args->backing_swap[0] = '\0';
		ON_ERR("backing_swap");
	}

	if (args->memlimit_kb) {
		VERBOSE("memlimit_kb: %" PRIu64 "\n", args->memlimit_kb);
		ret = ioctl(fd, RZSIO_SET_MEMLIMIT_KB, &args->memlimit_kb);
		args->memlimit_kb = 0;
		ON_ERR("memlimit_kb");
	}

	if (args->disksize_kb) {
		VERBOSE("disksize_kb: %" PRIu64 "\n", args->disksize_kb);
		ret = ioctl(fd, RZSIO_SET_DISKSIZE_KB, &args->disksize_kb);
		args->disksize_kb = 0;
		ON_ERR("disksize_kb");
	}

	if (args->init) {
		ret = ioctl(fd, RZSIO_INIT);
		args->init = 0;
		ON_ERR("init");
	}

	if (args->reset) {
		ret = ioctl(fd, RZSIO_RESET);
		args->reset = 0;
		ON_ERR("reset");
	}

	if (args->stats) {
		struct ramzswap_ioctl_stats s;
		memset(&s, 0, sizeof(s));
		ret = ioctl(fd, RZSIO_GET_STATS, &s);
		args->stats = 0;
		ON_ERR("stats");
		show_stats(&s);
	}
	
	} /* while (argc--) */

out:
	return ret;
}

int main(int argc, char *argv[])
{
	int fd, ret, opt, idx;
	struct rzs_args args;
	char dev_name[MAX_DEVICE_NAME_LEN];
	
	if (argc < 3) {
		usage();
		exit(0);
	}

	memset(&args, 0, sizeof(args));

	strncpy(dev_name, argv[1], MAX_DEVICE_NAME_LEN - 1);
	dev_name[MAX_DEVICE_NAME_LEN - 1] = '\0';

	fd = open(dev_name, O_NONBLOCK);
	if (fd == -1) {
		ret = errno;
		printf("Failed to open %s: %s\n", dev_name, strerror(ret));
		return ret;
	}

	VERBOSE("device: %s\n", dev_name);

	ret = 0;
	while (1) {
		char *endptr = NULL;

		static struct option long_options[] = {
			{ "backing_swap", required_argument, 0, 'b' },
			{ "memlimit_kb", required_argument, 0, 'm' },
			{ "disksize_kb", required_argument, 0, 'd' },
			{ "init", no_argument, 0, 'i' },
			{ "reset", no_argument, 0, 'r' },
			{ "stats", no_argument, 0, 's' },
			{ "verbose", no_argument, NULL, 'v' },
			{ "help", no_argument, NULL, 'h' },
			{ NULL, no_argument, NULL, 0 }
		};

		opt = getopt_long(argc, argv, "m:d:b:irsvh", long_options, &idx);

		if (opt == -1)
			break;
		
		switch (opt) {
		case 'b':
			strncpy(args.backing_swap, optarg, MAX_SWAP_NAME_LEN - 1);
			args.backing_swap[MAX_SWAP_NAME_LEN - 1] = '\0';
			break;

		case 'm':
			if (strnlen(optarg, MAX_SIZE_LEN + 1) > MAX_SIZE_LEN) {
				printf("memlimit_kb: %s\n", strerror(EOVERFLOW));
				ret = -EOVERFLOW;
				goto out;
			}

			args.memlimit_kb = strtoul(optarg, &endptr, 10);
			break;

		case 'd':
			if (strnlen(optarg, MAX_SIZE_LEN + 1) > MAX_SIZE_LEN) {
				printf("disksize_kb: %s\n", strerror(EOVERFLOW));
				ret = -EOVERFLOW;
				goto out;
			}

			args.disksize_kb = strtoul(optarg, &endptr, 10);
			break;

		case 's':
			args.stats = 1;
			break;

		case 'i':
			args.init = 1;
			break;

		case 'r':
			args.reset = 1;
			break;

		case 'v':
			verbose_flag = 1;
			break;

		case 'h':
			usage();
			break;

		case '?':
			usage();
			ret = -EINVAL;
			goto out;

		default:
			/* We never reach here */
			break;
		}
	}

	ret = do_ioctl(fd, argc - 2, &args);

out:
	close(fd);
	return ret;
}
